submitting-changes.chapter.md: add rules of thumb for master/staging … (#138052)
NixOS/nixpkgs: doc/contributing/submitting-changes.chapter.md

The `master` branch is the main development branch. It should only see non-breaking commits that do not cause mass rebuilds.

The guideline for proposed change is good for `master` branch:

> - Rebuild count is less than 1000 dependent 1000.

Reviewer: I don't understand that sentence.

Author: Oh wow. That sentence did not survive the edits /o\. Reworded the whole section and transformed into `Important update with rebuild count less than 1000 dependent packages and does not contain heavyweight packages like chromium or firefox. `

> - Rebuild count is between 1000 and 2000 lightweight packages (like python depends) and does not contain heavyweight package rebuilds (like `chromium` or `firefox`).

Reviewer: 2000 python rebuilds should probably also go to staging to not overwhelm Hydra. It is a rule of thumb. If the update is important it is easier to justify that it goes to master. If it is not important at all it should go to staging regardless.

Also it should be considered if the package update is likely to break depending packages. If so we should target staging to not break hundreds or thousands of packages.

Author: Decreased the cutoff limit to `1000` and mentioned simple leaf non-risky package bumps separately. Added more words on the intent not to overload Hydra.

- Kernel updates (they usually trigger about 2000 rebuilds).

> - Security fixes.

Reviewer: Also really depends on the severity. If the exploitation is very unlikely and not practical at all we don't want to rebuild the entire world. For example there are sometimes CVEs which need root to be exploitet which in my opinion do not justify rebuilds over 5000.

Author: Changed to somewhat vague `Urgent security fixes. `. People doing security fixes probably know the implications and finer details.

The `master` branch is the main development branch. It should only see non-breaking commits that do not cause mass rebuilds.

> What is a mass rebuild? It's an update that takes Hydra build infrastructure long time to process (say, 8 hours).

Reviewer: ```suggestion
What is a mass rebuild? It's an update that takes Hydra build infrastructure long time to process (a few days).
```

Here are rules of thumb that should help you decide if a change is reasonable for `master` branch:

> - Small update that rebuilds under 10 dependent packages unlikely to break users' systems. Examples are leaf package updates.

Reviewer: ```suggestion
- Small update that rebuilds under 100 dependent packages unlikely to break users' systems. Examples are leaf package updates.
```

> - Important update with rebuild count less than 1000 dependent packages and does not contain heavyweight packages like `chromium` or `firefox`.

Author: Tried to clarify with:
"""
- Important update of heavyweight leaf package like `chromium` or `firefox`
- Important update with rebuild count less than 1000 dependent packages. Batch close to 1000 should not contain heavyweight packages like `chromium` or `firefox`.
"""

> - Kernel updates.
> - Urgent security fixes.

Reviewer: ```suggestion
- Kernel updates
- Urgent security fixes
```

> Otherwise consider `staging` branch.

Reviewer: I am not sure if we should mention that stdenv or core programming language rebuilds like python need to go to staging.

Author: I think it's reasonable to provide a few "clearly for staging" examples. Added as

"""
+Otherwise consider `staging` branch. Typical changes for `staging` are:
+
+- Any changes to `stdenv` (trigger full rebuld)
+- Any core language ecosystem changes like `python` or `ghc` update (usually trigger more than 1000 dependent packages rebuild)
"""

The `master` branch is the main development branch. It should only see non-breaking commits that do not cause mass rebuilds.

What is a mass rebuild? It's an update that takes Hydra build infrastructure long time to process (a few days).

Here are rules of thumb that should help you decide if a change is reasonable for `master` branch:

- Small update that rebuilds under 100 dependent packages unlikely to break users' systems. Examples are leaf package updates.
- Important update of heavyweight leaf package like `chromium` or `firefox`
- Important update with rebuild count less than 1000 dependent packages. Batch close to 1000 should not contain heavyweight packages like `chromium` or `firefox`.
- Kernel updates
- Urgent security fixes

Otherwise consider `staging` branch. Typical changes for `staging` are:

- Any changes to `stdenv` (trigger full rebuld)

> - Any core language ecosystem changes like `python` or `ghc` update (usually trigger more than 1000 dependent packages rebuild)

Reviewer 2: `ghc` updates are done independently on `haskell-updates`, so a bad example. Large scale python changes are often done using `python-unstable`. I think language infrastructure is a bad example in general, since they are quite specific.

Better examples would be core libraries I think (gmp, ...).

Reviewer 2: `master`, but that's not really important here. Currently working on haskell documentation anyways and we'll just cover the haskell contributing quirks there, I think.

Reviewer: > ghc updates are done independently on haskell-updates, so a bad example.

Maybe in this case but they should be mentioned so that new contributors have a basic guiding.

Reviewer 2: I don't think it makes sense here — except if you want to overwhelm the reader. There are a lot of differences wrt workflow for certain subsystems of nixpkgs. If you want to make contributions to the haskell ecosystem, you'll want to read the (currently lacking) haskell contribution docs anyways. Besides, so far there have been zero occasions of someone PR-ing a GHC change on their first day…
